streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1227,24 +1227,30 @@ object ZSink extends ZSinkPlatformSpecificConstructors with Serializable {
   /**
    * Creates a sink accumulating incoming values into a map.
    * Key of each element is determined by supplied function.
+   * Combines elements with same key with supplied function f.
+   *
    */
-  final def collectAllToMap[K, A](key: A => K): ZSink[Any, Nothing, Nothing, A, Map[K, A]] =
-    foldLeft[A, Map[K, A]](Map.empty[K, A])((map, element) => map + (key(element) -> element))
+  final def collectAllToMap[K, A](key: A => K)(f: (A, A) => A): Sink[Nothing, Nothing, A, Map[K, A]] =
+    foldLeft[A, Map[K, A]](Map.empty)((curMap, a) => {
+      val k = key(a)
+      curMap.get(k).fold(curMap.updated(k, a))(v => curMap.updated(k, f(v, a)))
+    })
 
   /**
    * Creates a sink accumulating incoming values into a map of maximum size `n`.
    * Key of each element is determined by supplied function.
-   */
-  final def collectAllToMapN[K, A](n: Long)(key: A => K): Sink[Nothing, A, A, Map[K, A]] = {
-    type State = (Map[K, A], Boolean)
-    def f(state: State, a: A): (State, Chunk[A]) = {
-      val newMap = state._1 + (key(a) -> a)
-      if (newMap.size > n) ((state._1, false), Chunk.single(a))
-      else if (newMap.size == n) ((newMap, false), Chunk.empty)
-      else ((newMap, true), Chunk.empty)
-    }
-    fold[A, A, State]((Map.empty, true))(_._2)(f).map(_._1)
-  }
+   *
+   * Combines elements with same key with supplied function f.
+   */
+  final def collectAllToMapN[K, A](n: Long)(key: A => K)(f: (A, A) => A): Sink[Nothing, Nothing, A, Map[K, A]] =
+    foldLeft[A, Map[K, A]](Map.empty)((curMap, a) => {
+      val k = key(a)
+      curMap
+        .get(k)
+        .fold(
+          if (curMap.size >= n) curMap else curMap.updated(k, a)
```

Comment:
We're losing elements here. If we decide not to merge an element into the map, we need to push it back as a leftover.

Reply:
I wonder how I missed this. I thought it was a bit strange that there are no leftovers. :thinking:

Reply:
Yeah. The leftover was a bit confusing to me. Until I watched your talk on functional scala. Now I understand it! I went DOH when I read your comment. So, question now, How can I accumulate the leftovers? There is no run method that has that ability AFAIK. I would like to also make those reflected in tests. Also, I fail to understand why fold method has a Chunk[A] has the second type of tuple. Why not A and internally those would be accumulated into Chunk[A]? or Chunk[A0] rather? Sorry for the amount of questions. cc @iravid